scripts/image_customization_lib.lua

```diff
@@ -0,0 +1,132 @@
+local M = {}
+
+local function collect_keys(t)
+	local ret = {}
+	for v in pairs(t) do
+		table.insert(ret, v)
+	end
+	return ret
+end
+
+local function file_exists(file)
+	local f = io.open(file)
+	if not f then
+		return false
+	end
+	f:close()
+	return true
+end
+
+local function get_customization_file_name(env)
+	return env.GLUON_SITEDIR .. '/image-customization'
+end
+
+local function evaluate_device(env, dev)
+	local selections = {}
+	local funcs = {}
+	local device_overrides = {}
+
+	local function add_elements(element_type, element_list)
+		for _, element in ipairs(element_list) do
+			if not selections[element_type] then
+				selections[element_type] = {}
+			end
+
+			selections[element_type][element] = true
```

Comment:
The order of keys is unspecified, so this will break with negated elements. The elements will have to be stored as values, not as keys.